src/tarantool.client/LogicalConnection.cs

```diff
@@ -20,20 +17,97 @@ internal class LogicalConnection : ILogicalConnection
     {
         private readonly MsgPackContext _msgPackContext;
 
-        private readonly INetworkStreamPhysicalConnection _physicalConnection;
+        private readonly ClientOptions _clientOptions;
+
+        private readonly RequestIdCounter _requestIdCounter;
 
-        private long _currentRequestId;
+        private readonly INetworkStreamPhysicalConnection _physicalConnection;
 
-        private readonly ConcurrentDictionary<RequestId, TaskCompletionSource<MemoryStream>> _pendingRequests =
-            new ConcurrentDictionary<RequestId, TaskCompletionSource<MemoryStream>>();
+        private readonly IResponseReader _responseReader;
 
         private readonly ILog _logWriter;
 
-        public LogicalConnection(ClientOptions options, INetworkStreamPhysicalConnection physicalConnection)
+        private bool _disposed;
+
+        public LogicalConnection(ClientOptions options, RequestIdCounter requestIdCounter)
         {
+            _clientOptions = options;
+            _requestIdCounter = requestIdCounter;
             _msgPackContext = options.MsgPackContext;
             _logWriter = options.LogWriter;
-            _physicalConnection = physicalConnection;
+
+            _physicalConnection = new NetworkStreamPhysicalConnection();
+            _responseReader = new ResponseReader(_clientOptions, _physicalConnection);
+        }
+
+        public void Dispose()
+        {
+            lock (this)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                _disposed = true;
+
+                _responseReader.Dispose();
+                _physicalConnection.Dispose();
+            }
+        }
+
+        public async Task Connect()
+        {
+            await _physicalConnection.Connect(_clientOptions);
+
+            var greetingsResponseBytes = new byte[128];
+            var readCount = await _physicalConnection.ReadAsync(greetingsResponseBytes, 0, greetingsResponseBytes.Length);
+            if (readCount != greetingsResponseBytes.Length)
+            {
+                throw ExceptionHelper.UnexpectedGreetingBytesCount(readCount);
+            }
+
+            var greetings = new GreetingsResponse(greetingsResponseBytes);
+
+            _clientOptions.LogWriter?.WriteLine($"Greetings received, salt is {Convert.ToBase64String(greetings.Salt)} .");
+
+            _responseReader.BeginReading();
+
+            _clientOptions.LogWriter?.WriteLine("Server responses reading started.");
+
+            await LoginIfNotGuest(greetings);
+        }
+
+        public bool IsConnected()
+        {
+            if (_disposed)
+            {
+                return false;
```

Comment:
then we need catch exceptions here instead of check: https://github.com/progaudi/tarantool-csharp/pull/83/files/4be0d2ca63c07b90d6c7d9449e3da7ebb0624b65#diff-8b31d61323fab83274f6898980abbf6bR59